deltacat/catalog/v2/catalog_impl.py

```diff
@@ -0,0 +1,326 @@
+    LocalTable,
+)
+from deltacat.types.media import ContentType
+from deltacat.types.tables import TableWriteMode
+from deltacat.storage.rivulet.dataset import Dataset as RivuletDataset
+from deltacat.storage.main import impl as storage_impl
+from deltacat.storage.model.constants import (
+    DEFAULT_NAMESPACE,
+)
+
+
+# table functions
+def write_to_table(
+    data: Union[LocalTable, LocalDataset, DistributedDataset],
+    table: str,
+    namespace: Optional[str] = None,
+    mode: TableWriteMode = TableWriteMode.AUTO,
+    content_type: ContentType = ContentType.PARQUET,
+    *args,
+    **kwargs,
+) -> None:
+    raise NotImplementedError("Not implemented")
+
+
+def read_table(
+    table: str, namespace: Optional[str] = None, *args, **kwargs
+) -> DistributedDataset:
+    raise NotImplementedError("Not implemented")
+
+
+def alter_table(
+    table: str,
+    namespace: Optional[str] = None,
+    lifecycle_state: Optional[LifecycleState] = None,
+    schema_updates: Optional[Dict[str, Any]] = None,
+    partition_updates: Optional[Dict[str, Any]] = None,
+    sort_keys: Optional[SortScheme] = None,
+    description: Optional[str] = None,
+    properties: Optional[TableProperties] = None,
+    *args,
+    **kwargs,
+) -> None:
+    """Alter table definition."""
+    raise NotImplementedError("alter_table not implemented")
+
+
+def create_table(
+    table: str,
+    *args,
+    namespace: Optional[str] = None,
+    version: Optional[str] = None,
+    lifecycle_state: Optional[LifecycleState] = LifecycleState.ACTIVE,
+    schema: Optional[Schema] = None,
+    partition_scheme: Optional[PartitionScheme] = None,
+    sort_keys: Optional[SortScheme] = None,
+    description: Optional[str] = None,
+    table_properties: Optional[TableProperties] = None,
+    namespace_properties: Optional[NamespaceProperties] = None,
+    content_types: Optional[List[ContentType]] = None,
+    fail_if_exists: bool = True,
+    **kwargs,
+) -> TableDefinition:
+    """
+    Create an empty table. Raises an error if the table already exists and
+    `fail_if_exists` is True (default behavior).
+    """
+    catalog = kwargs.get("catalog")
+    if not isinstance(catalog, CatalogProperties):
+        raise ValueError("Catalog must be a CatalogProperties instance")
+
+    namespace = namespace or default_namespace()
+
+    # TODO
+    # Check if table exists
+    # if table_exists(table, namespace, **kwargs):
+    #     if fail_if_exists:
+    #         raise ValueError(f"Table {namespace}.{table} already exists")
+    #     return get_table(table, namespace, catalog=catalog)
+
+    # Create namespace if it doesn't exist
+    if not namespace_exists(namespace, catalog=catalog):
+        create_namespace(
+            namespace=namespace, properties=namespace_properties, catalog=catalog
+        )
+
+    # Extract merge keys from sort keys if provided
+    merge_keys = []
+    if sort_keys:
+        merge_keys = [key.name for key in sort_keys.keys]
```

Comment:
Hmm... do you always want Rivulet datasets to share the same merge and sort keys? Also it seems like we're either using the term `merge_keys` to mean slightly different things in different places, so we may want to use slightly different terminology to disambiguate (I think here they're effectively `zipper_merge_keys`, rather than the DeltaCAT schema `merge_keys` used to merge upserts/deletes by default - more similar to primary keys or Iceberg identifier field IDs).

Also, is Rivulet able to honor the given Sort Scheme (e.g., can it apply any optional transforms that may be attached to each sort key), or does it carry any preconceived notions about how the sort needs to occur that should be validated first?

Reply:
DeltaCAT's existing sort keys, similar to Iceberg sort/partition keys, allow you to define a transform that runs against each field prior to sorting or partitioning (e.g., to truncate a second-precision timestamp down to just day precision, or to produce the hash of a value, etc.).

I think we'll need to break the registration-only cases away from this as well, since we'll need to support that in addition to (or as part of) sorted Rivulet datasets.